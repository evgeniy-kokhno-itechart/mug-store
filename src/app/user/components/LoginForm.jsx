import React from 'react';
import _ from 'lodash';
import * as Yup from 'yup';
import { PropTypes } from 'prop-types';
import { FormBase, Input } from '../../shared';

export class LoginForm extends FormBase {
  state = {
    data: { username: '', password: '' },
    errors: {},
  };

  loginObjectSchema = {
    username: Yup.string()
      .required()
      .label('Username'),
    password: Yup.string()
      .required()
      .label('Password'),
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { data } = this.state;
    this.props.onSubmit(data.username, data.password);
  };

  render() {
    const { data, errors } = this.state;
    return (
      <React.Fragment>
        <h1 className="text-center m-3">Login</h1>
        <form className="col-10 col-md-8 col-lg-7 col-xl-5 mx-auto" onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="username"
            label="Username"
            value={_.get(data, 'username')}
            error={errors.username}
            validationSchema={this.loginObjectSchema.username}
            onChange={this.handleControlChange}
          />

          <Input
            type="password"
            name="password"
            label="Password"
            value={_.get(data, 'password')}
            error={errors.password}
            validationSchema={this.loginObjectSchema.password}
            onChange={this.handleControlChange}
          />

          <button type="submit" disabled={this.validateForm(this.loginObjectSchema, data)} className="btn btn-secondary w-100">
            Login
          </button>
        </form>
      </React.Fragment>
    );
  }
}

LoginForm.propTypes = {
  currentUserName: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
};

LoginForm.defaultProps = {
  currentUserName: '',
};

export default LoginForm;
