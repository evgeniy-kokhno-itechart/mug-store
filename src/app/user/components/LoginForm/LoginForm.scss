.login-form {
    &__header {
        text-align: center;
        margin: 1.5rem;
    }

    &__content {
        width: 50%;
        margin: 0 auto;
    }

    &-button {
        width: 100%;
        margin: 1rem 0;
    }
}
