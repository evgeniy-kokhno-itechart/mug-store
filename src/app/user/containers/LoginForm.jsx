import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import _ from 'lodash';
import * as Yup from 'yup';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { loginUser as loginUserAction } from '../UserActions';
import { FormService, Input } from '../../shared';

export class LoginForm extends Component {
  state = {
    data: { username: '', password: '' },
    errors: {},
  };

  loginObjectSchema = {
    username: Yup.string()
      .required()
      .label('Username'),
    password: Yup.string()
      .required()
      .label('Password'),
  };

  handleChange = (e, matchedInputName) => {
    const { target: input } = e;

    this.setState(prevState => FormService.handleChange(input, matchedInputName, prevState, this.loginObjectSchema));
  };

  handleSubmit = (e) => {
    e.preventDefault();
    try {
      const { data } = this.state;
      // eslint-disable-next-line no-shadow
      const { location, loginUserAction } = this.props;
      const { fromPath } = location;
      const redirectPath = fromPath ? fromPath.pathname : null;
      loginUserAction(data.username, data.password, redirectPath);
    } catch (ex) {
      const errors = {};
      errors.password = ex.message;
      this.setState({ errors });
    }
  };

  renderForm = () => {
    const { data, errors } = this.state;
    return (
      <React.Fragment>
        <h1 className="text-center m-3">Login</h1>
        <form className="col-10 col-md-8 col-lg-7 col-xl-5 mx-auto" onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="username"
            label="Username"
            value={_.get(data, 'username')}
            error={errors.username}
            onValueChange={this.handleChange}
          />

          <Input
            type="password"
            name="password"
            label="Password"
            value={_.get(data, 'password')}
            error={errors.password}
            onValueChange={this.handleChange}
          />

          <button type="submit" disabled={FormService.validateForm(this.loginObjectSchema, data)} className="btn btn-secondary w-100">
            Login
          </button>
        </form>
      </React.Fragment>
    );
  };

  render() {
    return this.props.currentUserName ? <Redirect to="/" /> : this.renderForm();
  }
}

LoginForm.propTypes = {
  currentUserName: PropTypes.string,
  location: PropTypes.shape({ fromPath: PropTypes.shape({ pathname: PropTypes.string }) }),
  loginUserAction: PropTypes.func.isRequired,
};

LoginForm.defaultProps = {
  currentUserName: '',
  location: { fromPath: null },
};

const mapStateToProps = state => ({ currentUserName: state.user.currentUser.name });

const mapDispatchToProps = {
  loginUserAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(LoginForm);
