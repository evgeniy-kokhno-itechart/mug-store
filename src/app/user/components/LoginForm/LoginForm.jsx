import React from 'react';
import _ from 'lodash';
import * as Yup from 'yup';
import { PropTypes } from 'prop-types';
import { FormBase, Input } from '../../../shared';
import './LoginForm.scss';

export class LoginForm extends FormBase {
  state = {
    data: { username: '', password: '' },
    errors: {},
  };

  loginObjectSchema = {
    username: Yup.string()
      .required()
      .label('Username'),
    password: Yup.string()
      .required()
      .label('Password'),
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { data } = this.state;
    this.props.onSubmit(data.username, data.password);
  };

  render() {
    const { data, errors } = this.state;
    return (
      <div className="login-form">
        <h1 className="login-form__header">Login</h1>
        <form className="login-form__content" onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="username"
            label="Username"
            value={_.get(data, 'username')}
            error={errors.username}
            validationSchema={this.loginObjectSchema.username}
            onChange={this.handleControlChange}
          />

          <Input
            type="password"
            name="password"
            label="Password"
            value={_.get(data, 'password')}
            error={errors.password}
            validationSchema={this.loginObjectSchema.password}
            onChange={this.handleControlChange}
          />

          <button
            type="submit"
            disabled={this.validateForm(this.loginObjectSchema, data)}
            className="button button--solid login-form-button"
          >
            Login
          </button>
        </form>
      </div>
    );
  }
}

LoginForm.propTypes = {
  currentUserName: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
};

LoginForm.defaultProps = {
  currentUserName: '',
};

export default LoginForm;
