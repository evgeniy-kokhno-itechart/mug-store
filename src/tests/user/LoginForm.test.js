/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { LoginForm } from '../../app/user/components/LoginForm/LoginForm';

configure({ adapter: new Adapter() });

describe('<LoginForm />', () => {
  let LoginFormWrapper;
  let onSubmit;

  beforeEach(() => {
    onSubmit = jest.fn();
    LoginFormWrapper = shallow(<LoginForm onSubmit={onSubmit} />);
  });

  it('renders properly with default props', () => {
    expect(LoginFormWrapper).toMatchSnapshot();
  });

  it('submits form properly after the from has been filled out', () => {
    const loginData = { username: 'testuser', password: 'testpass' };
    LoginFormWrapper.setState({ data: loginData });
    const form = LoginFormWrapper.find('form');
    form.simulate('submit', { preventDefault() {} });

    expect(onSubmit).toHaveBeenCalledWith(loginData.username, loginData.password);
  });
});
