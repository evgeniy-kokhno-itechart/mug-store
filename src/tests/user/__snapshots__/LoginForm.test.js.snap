// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> renders properly with default props 1`] = `
<div
  className="login-form"
>
  <h1
    className="login-form__header"
  >
    Login
  </h1>
  <form
    className="login-form__content"
    onSubmit={[Function]}
  >
    <Input
      error=""
      label="Username"
      name="username"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_label": "Username",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
    />
    <Input
      error=""
      label="Password"
      name="password"
      onChange={[Function]}
      type="password"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_label": "Password",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
    />
    <button
      className="button button--solid login-form-button"
      disabled={
        Object {
          "password": "Password is a required field",
          "username": "Username is a required field",
        }
      }
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
